### ISSUE TITLE
Need to make Web API endpoint stability clear

### ISSUE THREAD

**Issue by sheridancbio** [cBioPortal/cbioportal#1599]
The stability / commitment of various Web Api endpoints is not clear. The question is how should we communicate the "status" of the Web Api endpoints in a way to allow users to distinguish stable and supported endpoints from those which are under development or subject to change without notice.

@jjgao has proposed three categories of stability:
1. public and stable : available functionality will be available and maintained until major revision
2. in development : we think this endpoint may be added to "public and stable" in the future. For now, you can "use at your own risk" because unannounced changes are likely.
3. unstable : we do not think this endpoint will ever become stable. Deprecated and application specific endpoints will go here.

Is this a good categorization? Are we missing anything?

There are alternatives for how to communicate which category the endpoint falls into. The most visible way may be to put the category into the endpoint url path. That is the primary proposal here, that all endpoints fall under a path such as:
1. //{host}/{webapplicationname}/api/{endpoint} (for public and stable)
2. //{host}/{webapplicationname}/api-d/{endpoint} (for development / subject to change)
3. //{host}/{webapplicationname}/api-u/{endpoint} (for unstable / non-candidates to /api/)

We could have longer and more explicit names such as api-developing, api-develop, api-unstable, but sometimes shorter is better/cleaner. "beta" is not recommended because that is a testing process concept. Comments and alternatives are welcome. Hyphen versus underscore or other separator also is comment-worthy.

Information about the category of an endpoint could also be presented in other ways:
- wiki-style documentation pages
- explicit and universally provided swagger descriptive text tags
- through header fields in the api response itself
- through a "meta-api" which reported the stability, version, etc.. of each api endpoint

And a related question is: how should we handle versioning in the Web API? In the first public release should we require calling clients to specify the expected api version? Will the version requested be specified in:
- a parameter
- a content-type request header
- encoded in the URL path itself

The need for a quick consensus on this question is due to two new API endpoints which we want to make live soon without conflicting with previous endpoints, and with full knowledge that due to the Api Redesign efforts happening now, these two "in-development" endpoints will very likely need changes in the near future.

We would like to have an approach (at least a reasonable preliminary one) to follow soon. Target Date: Aug 23

**Comment by morungos** [comment]
Hmmm, not sure about moving the URL as they get developed. That's essentially a form of versioning, and RESTful APIs don't usually recommend that (even though most everybody does it). In particular, you probably don't want the unstable ones to be known, you can just not tell anybody what you're doing there. 

There's a nice set of discussions on this at: http://stackoverflow.com/questions/389169/best-practices-for-api-versioning. Needless to say, it's a flamey topic in the field, sadly. 

Since 90%+ of versioning is actually about representations, I like the use of application-specific content types to capture that.

However, the idea of defining endpoints that can, at any time, migrate from `/api-d` => `/api`, seems asking for trouble.

**Comment by inodb** [comment]
Would be cool if we can find an article or API implementation where we see this type of stable/develop/unstable URLs. To me it seems like a decent solution, but it's worrying that I don't see anybody else proposing something similar.

I don't think migrating endpoints is a huge problem. Those in `/api-d/` will only be used by us and a `git grep '/api-d/'` on the code base should make it trivial to rename them. The point would be to make it easy to add new endpoints for development w/o immediately having to release them. I think a bigger problem would be to add an endpoint and suddenly remove or significantly alter it in a new version of the API.

For backwards compatible changes e.g. simple representation changes or adding option parameters, we don't have to release a new version (http://searchsoa.techtarget.com/tip/Get-a-grip-How-to-handle-API-versioning-decisions).

**Comment by ersinciftci** [comment]
If we are going to use something like this as our API documentation: http://docs.cbioportal.apiary.io/, we can just not document it until it's stable. And if we deprecate something, we can mark it with **_(Deprecated)_** tag just like my current **_(Not Implemented Yet)_** tag. And about the versioning, I like this article https://www.troyhunt.com/your-api-versioning-is-wrong-which-is/, and I agree with the article and @morungos, I prefer using `Accept ` header rather than URL, something like `application/vnd.cbioportal.v2+json`. Used this in my previous company and worked well.

**Comment by ersinciftci** [comment]
Githup API uses `Accept` header too https://developer.github.com/v3/#current-version.

**Comment by inodb** [comment]
I'm happy with versioning in `Accept` headers. 

I'm not entirely sure how the REST API documentation/code is linked. I believe what @morungos and @ersinciftci  are proposing is: We use swagger to document the API, then generate the code for the API from the swagger yml, subsequently we add our own internal/development endpoints in the code. Front end developers know they have to look at the code to see full API. Once an endpoint is stable enough we add it to the swagger yml. Is that correct?

**Comment by ersinciftci** [comment]
I propose we use swagger yaml to document the API that we intend to have, but don't generate code, generated code is always ugly, just implement it using Spring MVC (we can generate code and use it as a guidance). And we won't generate the API doc from the code like we were doing until now, we will just visualize the swagger yaml file (Either with apiary like http://docs.cbioportal.apiary.io/ or with swagger-ui deployed on the cbioportal machine ). One might ask what happens if our API doc and implementation doesn't match for some cases, it means we have a bug in implementation and it's actually good that we don't have that bug in documentation so people can report it, so it's better to have them independent from each other. And for staging the API document to make things visible to frontend devs but not make it public, for example I have this http://docs.cbioersin.apiary.io synced to my fork of swagger spec, and I show this to people when my PR isn't merged yet. So we can use this beta kind of document approach. Or we can just make everything public and use http://docs.cbioportal.apiary.io/ all the time, but don't remove **_(Not Implemented Yet)_** tag until it's stable. And I don't really know the reason to have internal APIs, but if we have it, frontend devs can just ask backend devs about the endpoint since we don't want to have it in the public documentation.

**Comment by ersinciftci** [comment]
Or we can make swagger-ui and swagger yaml file part of the .war file, and swagger-ui can read swagger yaml file and generate the view instead of annotations in the code. So documentation staging becomes part of cbioportal's staging, /beta/swagger-ui.html can show APIs in beta. And internal APIs simply won't be part of the swagger yaml.

**Comment by n1zea144** [comment]
I think we should spend some time on Friday's call discussing some of these issues.  A related issue which we need to decide on soon is how to organize the Spring MVC code.  (if we do this, then that can inform where/how the documentation is generated).

We have a few new features under development here that require new endpoints (that have already been developed in the new tech stack) and we have to put them somewhere.  Also, what should happen to the existing ApiController that is currently supporting the Oncoprint?  It is in production right now, but we've never committed to it and it is probably going to chance or become obsolete in light of the mini-moonshot effort.  Should we move it?

**Comment by n1zea144** [comment]
To be clearer, I think having SpringFox document the public API via the Spring MVC package/location as part of the war/website deployment is the way to go.  I'm not so concerned about how we document the under development endpoints, but we have to decide where to place the code.

**Comment by ersinciftci** [comment]
I think we should organize the controller classes around the first word after `/api/`, like StudyController, SampleController etc. so classes can have `@RequestMapping("/studies")`, `@RequestMapping("/samples")` in the class level, so the methods don't need to repeat it. As for ApiController or other APIs independent from mini-moonshot, we better don't mix them with the mini-moonshot ones. So new non-minimoonshot ones can still go under ApiController.

**Comment by ersinciftci** [comment]
If we are going to continue to use SpringFox annotation documentation way, I won't invest more time in https://github.com/cBioPortal/cbioportal-web-api-spec(which is already 700 lines) because it doesn't make sense to create this spec and then generate documentation from the code instead of this spec.

**Comment by n1zea144** [comment]
..but we do need some way to discuss the spec without having to write MVC controllers.   Lets discuss  more on Friday.

**Comment by ersinciftci** [comment]
Exactly, and we can reuse that spec to create the same swagger-ui documentation instead of SpringFox annotations. ApiController looks ugly because of all those @ApiOperation and @ApiParam annotations. Yeah lets discuss more on Friday, I'm continuing investing time then :)

**Comment by sheridancbio** [comment]
Thanks for the good points. I think I also have come to the viewpoint that we should not try to use the API URL pattern to communicate what is essentially meta information about the api endpoints (Their stability and the degree we commit to maintaining them .. or the release version info) Moving endpoints from /api-d/ to /api/ does have some overhead for us --- and might mess up third parties too. I share Stuart's sense that there is probably other costs we are unaware of now which will come back to bite us later. Still interested in hearing any other thoughts on this though. I also really liked the article from troyhunt.com that Ersin referenced .. and the idea of using the Accept header seems good.

I personally like the idea of being more documentation / specification driven, but I worry about moving away from auto-generation of swagger spec using springfox. One good thing about embedded annotations and auto-generation is that it forces you (or at least seriously encourages you) to update the documentation when you make changes to the code. For a published and stable API endpoint, there should not be much spontaneous tweaking of the arguments, return type, and descriptions. But for the development API endpoints there will be tweaks, and we do want the development API to be visible and documented --- even if it is not committed yet to be stable we still want to be able to pull it up and see it. I suspect that if we don't have the annotations right there in the code, we won't always remember to get to the "update the documentation" step of the task list.

**Comment by ersinciftci** [comment]
Actually we can do something like this `.apis(not(withClassAnnotation(NotStable.class))` in our `ApiServiceSwaggerConfig` swagger configuration, so the methods annotated with our custom annotation `@NotStable` won't be in the swagger-ui, and remove the annotation when it's stable, instead of moving between packages.

And actually the spec I'm creating is good for mocking too even though if we are not going to use it for documentation, so it's not a bad investment even if we are going to use SpringFox. I forgot that detail, sorry :)

**Comment by ersinciftci** [comment]
I agree @sheridancbio about not hiding anything actually. If something is not ready yet, it won't it make to master branch anyway, so it will be under `/beta/api/swagger-ui.html`, which is fine. And internal APIs are already in our open source repos, so we can't really hide. "Ye be warned" should be enough to keep people away if we want to.

**Comment by ersinciftci** [comment]
I thought more on this and I think if we are going to use SpringFox, instead of creating swagger yaml file to show people the API design and get feedback, I can just go ahead and create SpringMVC controllers with Swagger annotations, and methods would return [UnsupportedOperationException](https://docs.oracle.com/javase/8/docs/api/java/lang/UnsupportedOperationException.html) until we implement them, and SpringFox would still generate the docs for them. It would take the same or less time than writing the spec, and it's something we are going to write anyways. And if we change `ApiServiceSwaggerConfig` like this:

```
@Configuration
@EnableSwagger2
public class ApiServiceSwaggerConfig {
    @Bean
    public Docket api(){
        return new Docket(DocumentationType.SWAGGER_2)
            .select()
            .apis(RequestHandlerSelectors.basePackage("org.mskcc.cbio.portal.web.api"))
            .build()
            .pathMapping("/api")
            .apiInfo(apiInfo());
    }

    @Bean
    public Docket design(){
        return new Docket(DocumentationType.SWAGGER_2).groupName("design")
                .select()
                .apis(RequestHandlerSelectors.basePackage("org.cbioportal.web"))
                .build()
                .pathMapping("/api")
                .apiInfo(apiInfo());
    }

    private ApiInfo apiInfo() {
        ApiInfo apiInfo = new ApiInfo(
            "cBioPortal web API", 
            "A web service for supplying JSON formatted data to cBioPortal clients.", 
            "1.0 (beta)",
            "www.cbioportal.org",
            "cbioportal@googlegroups.com", 
            "License", 
            "https://github.com/cBioPortal/cbioportal/blob/master/LICENSE");
        return apiInfo;
    }
}
```

the dropdown on the top of the swagger-ui page will have 2 values, first is the `default`, current `api-controller`, and the second is `design`, showing the current `mutation-controller` under the `org.cbioportal.web` package. And mocks can be crated manually from the response model schema in swagger-ui. And in this way we can have `.groupName("internal")` too.

So tomorrow I can start writing controllers that returns `UnsupportedOperationException` for the APIs in the current yaml file and continue new APIs from there, if everybody agrees.

**Comment by n1zea144** [comment]
Does swagger-yaml and SpringFox have to be mutually exclusive?  Is there something wrong with using apiary to discuss/develop api design and then during controller implementation, we mark up with  SpringFox annotations too?  Thats an extra level of sanity checking and really just part of the implementation.

**Comment by ersinciftci** [comment]
They are two alternative concepts, contract-first vs code-first, you either generate the code from the contract, or generate the contract from the code, if you create both of them manually, it takes more time and they can very easily go out of sync. Swagger-yaml is not a documentation markup language, you create a fully featured APIs with your models and controllers using yaml, instead of java, it just doesn't have the actual service layer code. Apiary is using that file to generate documentation, just like SpringFox use your java code. So using swagger-yaml+apiary just for design/discuss purposes is like for example creating a dummy API in Python+Flask+Flasgger(a library like SpringFox) first to generate a documentation to discuss, and then implementing it in java+SpringFox again, then when you want to change something, first change in the Python app, get feedback then implement in the java app. So I'm proposing, why not go ahead and add the dummy APIs in java+SpringFox for design/discuss purposes, then when it comes to implementation, just delete `UnsupportedOperationException` and call the service layer method instead.

**Comment by ersinciftci** [comment]
Creating swagger-yaml and using `https://github.com/garethjevans/swagger-codegen-maven-plugin` to generate the Spring MVC controllers and domain models from yaml file is another option, but not as flexible as code-first approach, and generated code is kind of ugly.

**Comment by sheridancbio** [comment]
Maybe we can talk about this question during the call tomorrow a little bit. I'll try to put together a summary document so we can be efficient.

One thought I have about UnsupportedOperationException ... I think that is fine for the stage where we are brainstorming API proposals ... effectively using springfox as a chalkboard for our API concepts (If I understand this correctly). But then there will come a time when we want to start actually using the API endpoint in production code because we have built a preliminary react page frontend and a preliminary backend persistence / service stack to provide the API functionality. When we want to "turn on" the endpoint for the first time, at that time we probably will not be ready to commit to a long term contract to support the released interface. I think we will want the freedom to fine tune or adjust the endpoint for a while to deal with any issues that are discovered after deployment.  So I think we would need to take away the UnsupportedOperationException but still be in a "development" state. But maybe that is where the annotation @NotStable comes in .. we leave that on until we are ready to make a public commitment of stability. I'll try to capture these ideas clearly.

**Comment by ersinciftci** [comment]
I thought more on versioning, I think it depends if we are going to have different deployments for different versions, or the same war serves all the versions. If we are going to have multiple deployments, we can deploy them under different paths, like `/v1/` and `/v2/`, so `Accept` header doesn't make in this case, if we are going to serve all the versions from the same application, then `Accept` header makes sense. I think it's easer to have multiple deployments for multiple versions, otherwise we may need things like `if (version == 2)` in the code, which is ugly.
